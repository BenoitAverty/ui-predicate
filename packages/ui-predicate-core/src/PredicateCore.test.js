// why? because I HATE ES6 import/export
const {
  PredicateCore,
  dataclasses: {
    Target,
    Operator,
    Predicates,
    CompoundPredicate,
    ComparisonPredicate,
    LogicalType,
  },
  errors,
} = require('..');

const { prop } = require('ramda');

describe('core.component', () => {
  describe('constructor', () => {
    it('throws an error if data does not start with a ComparisonPredicate', () => {
      expect(() => new PredicateCore({ data: [] })).toThrow();
    });

    it('default state (without passed data) should have one predicate, selected automatically', () => {
      const ctrl = new PredicateCore();
      expect(ctrl.root.predicates[0].target.$type).toBeDefined();
      expect(ctrl.root.predicates[0].target.$type.$operators).toBeDefined();
      expect(ctrl.toJSON()).toMatchSnapshot();
    });

    it('check that every targets refer to a defined type', () => {
      expect(
        () =>
          new PredicateCore({
            columns: {
              // besides array list names, everything else follows convention https://github.com/FGRibreau/sql-convention
              operators: [],
              types: [],
              targets: [
                {
                  target_id: 'article.title',
                  label: 'Titre article',
                  type_id: 'string',
                },
              ],
            },
          })
      ).toThrow(errors.TargetMustReferToADefinedType);
    });
  });

  describe('ctrl.add', () => {
    it('throws if Predicate type is invalid', () => {
      const ctrl = new PredicateCore();
      const firstPredicate = ctrl.root.predicates[0];
      expect(() =>
        ctrl.add({
          where: firstPredicate,
          type: 'Blablabla',
        })
      ).toThrow(errors.InvalidPredicateType);
    });

    it('throws if how is unsupported', () => {
      const ctrl = new PredicateCore();
      const firstPredicate = ctrl.root.predicates[0];
      expect(() =>
        ctrl.add({
          where: firstPredicate,
          how: 'insteadof',
          type: 'CompoundPredicate',
        })
      ).toThrow(errors.AddOnlySupportsAfter);
    });

    it('adds a Predicate (a CompoundPredicate) after the first predicate to the root CompoundPredicate - without how parameter', () => {
      const ctrl = new PredicateCore();
      const firstPredicate = ctrl.root.predicates[0];
      ctrl.add({
        where: firstPredicate,
        type: 'CompoundPredicate',
      });

      expect(ctrl.root.predicates.length).toEqual(2);
      expect(CompoundPredicate.is(ctrl.root.predicates[1])).toBe(true);
    });

    it('adds a Predicate (a CompoundPredicate) after the first predicate of the root CompoundPredicate', () => {
      const ctrl = new PredicateCore();
      const firstPredicate = ctrl.root.predicates[0];
      ctrl.add({
        where: firstPredicate,
        how: 'after',
        type: 'CompoundPredicate',
      });

      expect(ctrl.root.predicates.length).toEqual(2);
      expect(CompoundPredicate.is(ctrl.root.predicates[1])).toBe(true);
    });

    it('adds a Predicate (a ComparisonPredicate) after the first predicate of the root CompoundPredicate', () => {
      const ctrl = new PredicateCore();
      const firstPredicate = ctrl.root.predicates[0];
      ctrl.add({
        where: firstPredicate,
        how: 'after',
        type: 'ComparisonPredicate',
      });

      expect(ctrl.root.predicates.length).toEqual(2);
      expect(ComparisonPredicate.is(ctrl.root.predicates[1])).toBe(true);
    });

    it('(currently unsupported) add a Predicate (a ComparisonPredicate) after the CompoundPredicate', () => {
      const ctrl = new PredicateCore();
      const firstPredicate = ctrl.root.predicates[0];
      const compoundPredicate = ctrl.add({
        where: firstPredicate,
        how: 'after',
        type: 'CompoundPredicate',
      });

      expect(CompoundPredicate.is(compoundPredicate)).toBe(true);

      expect(() =>
        ctrl.add({
          where: compoundPredicate,
          how: 'after',
          type: 'ComparisonPredicate',
        })
      ).toThrow('unsupported');
    });

    // it('adds a second Predicate (a ComparisonPredicate) to the root CompoundPredicate', () => {
    //   const ctrl = new PredicateCore();
    //   const firstPredicate = ctrl.root.predicates[0];
    //   ctrl.add({ after: firstPredicate });
    //   expect(ctrl.toJSON()).toMatchSnapshot();
    // });
  });

  describe('ctrl.setPredicateTarget_id', () => {
    it('throws when predicate refers to a invalid predicate', () => {
      const ctrl = new PredicateCore();
      const firstPredicate = ctrl.root.predicates[0];
      expect(() =>
        ctrl.setPredicateTarget_id(null, ctrl.columns.targets[1].target_id)
      ).toThrow(errors.PredicateMustBeAComparisonPredicate);
    });

    it('throws when target_id refers to a not defined target', () => {
      const ctrl = new PredicateCore();
      const firstPredicate = ctrl.root.predicates[0];
      expect(() =>
        ctrl.setPredicateTarget_id(firstPredicate, 'blablablabla')
      ).toThrow(errors.Target_idMustReferToADefinedTarget);
    });

    it('work if we set a predicate to an exist target_id', () => {
      const ctrl = new PredicateCore();
      const firstPredicate = ctrl.root.predicates[0];

      expect(firstPredicate.target.target_id).not.toBe(
        ctrl.columns.targets[1].target_id
      );

      expect(firstPredicate.operator.operator_id).not.toBe(
        ctrl.columns.targets[1].$type.$operators[0].operator_id
      );

      ctrl.setPredicateTarget_id(
        firstPredicate,
        ctrl.columns.targets[1].target_id
      );

      // should update target
      expect(firstPredicate.target.target_id).toBe(
        ctrl.columns.targets[1].target_id
      );

      // should update operator as well
      expect(firstPredicate.operator.operator_id).toBe(
        ctrl.columns.targets[1].$type.$operators[0].operator_id
      );

      expect(firstPredicate.arguments).toEqual([]);
      expect(firstPredicate).toMatchSnapshot();
    });
  });

  describe('ctrl.setPredicateOperator_id', () => {
    it('throws an error if operator_id cannot be found', () => {
      const ctrl = new PredicateCore();
      const firstPredicate = ctrl.root.predicates[0];
      expect(() =>
        ctrl.setPredicateOperator_id(firstPredicate, 'lool')
      ).toThrow(errors.Operator_idMustReferToADefinedOperator);
    });

    it('work if we set a predicate to an existing operator_id', () => {
      const ctrl = new PredicateCore();
      const firstPredicate = ctrl.root.predicates[0];
      const target_idBefore = firstPredicate.target.target_id;

      expect(firstPredicate.operator.operator_id).not.toBe(
        firstPredicate.target.$type.$operators[1].operator_id
      );

      ctrl.setPredicateOperator_id(
        firstPredicate,
        firstPredicate.target.$type.$operators[1].operator_id
      );

      // should not update target
      expect(firstPredicate.target.target_id).toBe(target_idBefore);

      // should only update operator
      expect(firstPredicate.operator.operator_id).toBe(
        firstPredicate.target.$type.$operators[1].operator_id
      );
    });
  });

  describe('ctrl.options', () => {
    it('expose computed options', () => {
      const ctrl = new PredicateCore();
      expect(Object.keys(ctrl.options)).toEqual([
        'getDefaultData',
        'getDefaultCompoundPredicate',
        'getDefaultComparisonPredicate',
      ]);
    });
  });

  // describe('ctrl.addCompoundPredicate({after:Predicate})', () => {}

  // describe('ctrl.remove(predicate)', () => {
  //   it('forbids to remove the last predicate of the root CompoundPredicate', () => {
  //     const ctrl = new PredicateCore();
  //     const firstPredicate = ctrl.root.predicates[0];
  //     ctrl.remove(firstPredicate);
  //   });
  // });
});

describe('core.data', () => {
  describe('CompoundPredicate', () => {
    describe('constructor', () => {
      it(`can't be constructed with at least one subpredicate`, () => {
        expect(() => CompoundPredicate(LogicalType.and, [])).toThrow();
      });
    });
  });

  // describe('serializable', () => {
  //   it('serialize a CompoundPredicate with 2 subpredicates', () => {
  //     const root = CompoundPredicate(LogicalType.and, [
  //       ComparisonPredicate(Target(), Operator(), []),
  //       ComparisonPredicate(Target(), Operator(), []),
  //     ]);
  //
  //     expect(JSON.stringify(root)).toMatchSnapshot();
  //   });
  //
  //   it('serialize a CompoundPredicate with subpredicates composed of CompoundPredicate', () => {
  //     const root = CompoundPredicate(LogicalType.and, [
  //       ComparisonPredicate(Target(), Operator(), []),
  //       ComparisonPredicate(Target(), Operator(), []),
  //       CompoundPredicate(LogicalType.and, [
  //         ComparisonPredicate(Target(), Operator(), []),
  //       ]),
  //     ]);
  //
  //     expect(JSON.stringify(root, null, 2)).toMatchSnapshot();
  //   });
  // });

  // describe(`Root CompoundPredicate can't have less (and it child) than 1 Predicate`, () => {
  //   it('works', () => {
  //     const sub1 = ComparisonPredicate(Target(), Operator(), []);
  //     const sub2 = ComparisonPredicate(Target(), Operator(), []);
  //     const root = CompoundPredicate(LogicalType.and, [sub1, sub2]);
  //
  //     root.remove(sub2);
  //     return sub2.remove().then(() => sub.remove());
  //   });
  // });
});
