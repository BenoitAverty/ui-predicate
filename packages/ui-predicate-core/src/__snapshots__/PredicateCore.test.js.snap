// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UIPredicateCore CompoundPredicate constructor can't be constructed with at least one subpredicate 1`] = `[CompoundPredicateMustHaveAtLeastOneSubPredicate]`;

exports[`UIPredicateCore constructor check that every targets refer to a defined type 1`] = `[TargetMustReferToADefinedType: target "article.title" does not refer to a defined type, target.type_id="string"]`;

exports[`UIPredicateCore constructor default state (without passed data) should have one predicate, selected automatically 1`] = `
Object {
  "logicalType_id": "any",
  "predicates": Array [
    Object {
      "argument": null,
      "operator_id": "is",
      "target_id": "title",
    },
  ],
}
`;

exports[`UIPredicateCore constructor expose computed options 1`] = `
Array [
  "on",
  "once",
  "off",
  "setData",
  "add",
  "remove",
  "setPredicateTarget_id",
  "setPredicateOperator_id",
  "setPredicateLogicalType_id",
  "setArgumentValue",
  "getArgumentTypeComponentById",
  "getUIComponent",
  "toJSON",
  "root",
  "columns",
  "options",
]
`;

exports[`UIPredicateCore constructor rejects an error if json data does not start with a ComparisonPredicate 1`] = `[RootPredicateMustBeACompoundPredicate]`;

exports[`UIPredicateCore constructor rejects an error if json data is neither a CompoundPredicate or a ComparisonPredicate 1`] = `[UnknownJSONData]`;

exports[`UIPredicateCore constructor throws columns is not an object 1`] = `[Error: Object is required, got: null.]`;

exports[`UIPredicateCore constructor throws if columns does not have minimum required parameters 1`] = `[Error: Object {} MUST have a 'targets' property.]`;

exports[`UIPredicateCore constructor throws if options.getDefaultArgumentComponent is not overriden by the UI Framework adapter 1`] = `"UIFrameworkMustImplementgetDefaultArgumentComponent"`;

exports[`UIPredicateCore ctrl.add adds a second Predicate (a ComparisonPredicate) to the root CompoundPredicate 1`] = `
Object {
  "logicalType_id": "any",
  "predicates": Array [
    Object {
      "argument": null,
      "operator_id": "is",
      "target_id": "title",
    },
    Object {
      "argument": null,
      "operator_id": "is",
      "target_id": "title",
    },
  ],
}
`;

exports[`UIPredicateCore ctrl.add rejects if \`how\` is unsupported 1`] = `[AddCurrentlyOnlySupportAfterInsertion]`;

exports[`UIPredicateCore ctrl.add rejects if Predicate type is invalid 1`] = `[InvalidPredicateType]`;

exports[`UIPredicateCore ctrl.add rejects if where is unsupported 1`] = `[CannotAddSomethingElseThanACompoundPredicateOrAComparisonPredicate]`;

exports[`UIPredicateCore ctrl.remove(predicate) allow to remove a comparisonPredicate 1`] = `
Object {
  "$_type": "CompoundPredicate",
  "$canBeRemoved": false,
  "logic": Object {
    "$_type": "LogicalType",
    "label": "Any",
    "logicalType_id": "any",
  },
  "predicates": Array [
    Object {
      "$_type": "ComparisonPredicate",
      "$canBeRemoved": false,
      "argument": null,
      "operator": Object {
        "$_type": "Operator",
        "argumentType_id": "smallString",
        "label": "is",
        "operator_id": "is",
      },
      "target": Object {
        "$_type": "Target",
        "$type": Object {
          "$_type": "Type",
          "$operators": Array [
            Object {
              "$_type": "Operator",
              "argumentType_id": "smallString",
              "label": "is",
              "operator_id": "is",
            },
            Object {
              "$_type": "Operator",
              "argumentType_id": "smallString",
              "label": "Contains",
              "operator_id": "contains",
            },
          ],
          "operator_ids": Array [
            "is",
            "contains",
          ],
          "type_id": "string",
        },
        "label": "Title",
        "target_id": "title",
        "type_id": "string",
      },
    },
  ],
}
`;

exports[`UIPredicateCore ctrl.remove(predicate) allow to remove a comparisonPredicate and its parent CompoundPredicate if it was the last comparisonPredicate 1`] = `
Object {
  "$_type": "CompoundPredicate",
  "$canBeRemoved": false,
  "logic": Object {
    "$_type": "LogicalType",
    "label": "Any",
    "logicalType_id": "any",
  },
  "predicates": Array [
    Object {
      "$_type": "ComparisonPredicate",
      "$canBeRemoved": false,
      "argument": null,
      "operator": Object {
        "$_type": "Operator",
        "argumentType_id": "smallString",
        "label": "is",
        "operator_id": "is",
      },
      "target": Object {
        "$_type": "Target",
        "$type": Object {
          "$_type": "Type",
          "$operators": Array [
            Object {
              "$_type": "Operator",
              "argumentType_id": "smallString",
              "label": "is",
              "operator_id": "is",
            },
            Object {
              "$_type": "Operator",
              "argumentType_id": "smallString",
              "label": "Contains",
              "operator_id": "contains",
            },
          ],
          "operator_ids": Array [
            "is",
            "contains",
          ],
          "type_id": "string",
        },
        "label": "Title",
        "target_id": "title",
        "type_id": "string",
      },
    },
  ],
}
`;

exports[`UIPredicateCore ctrl.remove(predicate) allow to remove a compoundpredicate 1`] = `
Object {
  "$_type": "CompoundPredicate",
  "$canBeRemoved": true,
  "logic": Object {
    "$_type": "LogicalType",
    "label": "Any",
    "logicalType_id": "any",
  },
  "predicates": Array [
    Object {
      "$_type": "ComparisonPredicate",
      "$canBeRemoved": true,
      "argument": null,
      "operator": Object {
        "$_type": "Operator",
        "argumentType_id": "smallString",
        "label": "is",
        "operator_id": "is",
      },
      "target": Object {
        "$_type": "Target",
        "$type": Object {
          "$_type": "Type",
          "$operators": Array [
            Object {
              "$_type": "Operator",
              "argumentType_id": "smallString",
              "label": "is",
              "operator_id": "is",
            },
            Object {
              "$_type": "Operator",
              "argumentType_id": "smallString",
              "label": "Contains",
              "operator_id": "contains",
            },
          ],
          "operator_ids": Array [
            "is",
            "contains",
          ],
          "type_id": "string",
        },
        "label": "Title",
        "target_id": "title",
        "type_id": "string",
      },
    },
  ],
}
`;

exports[`UIPredicateCore ctrl.remove(predicate) forbids to remove an unknown type of predicate 1`] = `[CannotRemoveSomethingElseThanACompoundPredicateOrAComparisonPredicate]`;

exports[`UIPredicateCore ctrl.remove(predicate) forbids to remove the last predicate of the root CompoundPredicate 1`] = `[ForbiddenCannotRemoveLastComparisonPredicate]`;

exports[`UIPredicateCore ctrl.remove(predicate) forbids to remove the root CompoundPredicate 1`] = `[ForbiddenCannotRemoveRootCompoundPredicate]`;

exports[`UIPredicateCore ctrl.setPredicateLogicalType_id yield a rejected promise if logicalType_id cannot be found 1`] = `[LogicalType_idMustReferToADefinedLogicalType]`;

exports[`UIPredicateCore ctrl.setPredicateLogicalType_id yield a rejected promise if predicate is not a CompoundPredicate 1`] = `[PredicateMustBeACompoundPredicate]`;

exports[`UIPredicateCore ctrl.setPredicateOperator_id rejects an error if operator_id cannot be found 1`] = `[Operator_idMustReferToADefinedOperator]`;

exports[`UIPredicateCore ctrl.setPredicateTarget_id rejects when predicate refers to a invalid predicate 1`] = `[PredicateMustBeAComparisonPredicate]`;

exports[`UIPredicateCore ctrl.setPredicateTarget_id rejects when target_id refers to a not defined target 1`] = `[Target_idMustReferToADefinedTarget]`;

exports[`UIPredicateCore ctrl.setPredicateTarget_id work if we set a predicate to an exist target_id 1`] = `
Object {
  "$_type": "ComparisonPredicate",
  "$canBeRemoved": false,
  "argument": null,
  "operator": Object {
    "$_type": "Operator",
    "argumentType_id": "number",
    "label": "<",
    "operator_id": "isLowerThan",
  },
  "target": Object {
    "$_type": "Target",
    "$type": Object {
      "$_type": "Type",
      "$operators": Array [
        Object {
          "$_type": "Operator",
          "argumentType_id": "number",
          "label": "<",
          "operator_id": "isLowerThan",
        },
        Object {
          "$_type": "Operator",
          "argumentType_id": "number",
          "label": "=",
          "operator_id": "isEqualTo",
        },
        Object {
          "$_type": "Operator",
          "argumentType_id": "number",
          "label": ">",
          "operator_id": "isHigherThan",
        },
      ],
      "operator_ids": Array [
        "isLowerThan",
        "isEqualTo",
        "isHigherThan",
      ],
      "type_id": "int",
    },
    "label": "Video count",
    "target_id": "videoCount",
    "type_id": "int",
  },
}
`;

exports[`UIPredicateCore toJSON serialize without special flags 1`] = `
Object {
  "logicalType_id": "any",
  "predicates": Array [
    Object {
      "argument": null,
      "operator_id": "is",
      "target_id": "title",
    },
    Object {
      "logicalType_id": "any",
      "predicates": Array [
        Object {
          "argument": null,
          "operator_id": "is",
          "target_id": "title",
        },
        Object {
          "logicalType_id": "any",
          "predicates": Array [
            Object {
              "argument": null,
              "operator_id": "is",
              "target_id": "title",
            },
          ],
        },
      ],
    },
  ],
}
`;
